--- a/net/minecraft/server/BlockStateBoolean.java
+++ b/net/minecraft/server/BlockStateBoolean.java
@@ -5,12 +5,31 @@
 
 public class BlockStateBoolean extends BlockState<Boolean> {
 
-    private final ImmutableSet<Boolean> a = ImmutableSet.of(true, false);
+    private final ImmutableSet<Boolean> a = ImmutableSet.of(Boolean.valueOf(true), Boolean.valueOf(false));
 
     protected BlockStateBoolean(String s) {
         super(s, Boolean.class);
     }
 
+    // TacoSpigot start
+    @Override
+    public int getValueId(Boolean value) {
+        return value ? 1 : 0;
+    }
+
+    @Override
+    public Boolean getByValueId(int id) {
+        switch (id) {
+            case 0:
+                return false;
+            case 1:
+                return true;
+            default:
+                throw new IllegalArgumentException("Invalid id: " + id);
+        }
+    }
+    // TacoSpigot end
+
     public Collection<Boolean> c() {
         return this.a;
     }
@@ -23,7 +42,11 @@
         return obool.toString();
     }
 
+    // TacoSpigot start - fix stupid generic thingies
+    /*
     public String a(Comparable comparable) {
         return this.a((Boolean) comparable);
     }
+    */
+    // CraftBukkit end
 }
