--- a/net/minecraft/server/BlockDaylightDetector.java
+++ b/net/minecraft/server/BlockDaylightDetector.java
@@ -10,7 +10,7 @@
     public BlockDaylightDetector(boolean flag) {
         super(Material.WOOD);
         this.b = flag;
-        this.j(this.blockStateList.getBlockData().set(BlockDaylightDetector.POWER, 0));
+        this.j(this.blockStateList.getBlockData().set(BlockDaylightDetector.POWER, Integer.valueOf(0)));
         this.a(0.0F, 0.0F, 0.0F, 1.0F, 0.375F, 1.0F);
         this.a(CreativeModeTab.d);
         this.c(0.2F);
@@ -23,7 +23,7 @@
     }
 
     public int a(IBlockAccess iblockaccess, BlockPosition blockposition, IBlockData iblockdata, EnumDirection enumdirection) {
-        return (Integer) iblockdata.get(BlockDaylightDetector.POWER);
+        return ((Integer) iblockdata.get(BlockDaylightDetector.POWER)).intValue();
     }
 
     public void f(World world, BlockPosition blockposition) {
@@ -40,8 +40,9 @@
                 i = 15 - i;
             }
 
-            if ((Integer) iblockdata.get(BlockDaylightDetector.POWER) != i) {
-                world.setTypeAndData(blockposition, iblockdata.set(BlockDaylightDetector.POWER, i), 3);
+            if (((Integer) iblockdata.get(BlockDaylightDetector.POWER)).intValue() != i) {
+                i = org.bukkit.craftbukkit.event.CraftEventFactory.callRedstoneChange(world, blockposition.getX(), blockposition.getY(), blockposition.getZ(), ((Integer) iblockdata.get(POWER)), i).getNewCurrent(); // CraftBukkit - Call BlockRedstoneEvent
+                world.setTypeAndData(blockposition, iblockdata.set(BlockDaylightDetector.POWER, Integer.valueOf(i)), 3);
             }
 
         }
@@ -92,14 +93,14 @@
     }
 
     public IBlockData fromLegacyData(int i) {
-        return this.getBlockData().set(BlockDaylightDetector.POWER, i);
+        return this.getBlockData().set(BlockDaylightDetector.POWER, Integer.valueOf(i));
     }
 
     public int toLegacyData(IBlockData iblockdata) {
-        return (Integer) iblockdata.get(BlockDaylightDetector.POWER);
+        return ((Integer) iblockdata.get(BlockDaylightDetector.POWER)).intValue();
     }
 
     protected BlockStateList getStateList() {
-        return new BlockStateList(this, new IBlockState[]{BlockDaylightDetector.POWER});
+        return new BlockStateList(this, new IBlockState[] { BlockDaylightDetector.POWER});
     }
 }
