--- a/net/minecraft/server/ChatBaseComponent.java
+++ b/net/minecraft/server/ChatBaseComponent.java
@@ -56,7 +56,7 @@
     }
 
     public Iterator<IChatBaseComponent> iterator() {
-        return Iterators.concat(Iterators.forArray(new ChatBaseComponent[]{this}), a((Iterable) this.a));
+        return Iterators.concat(Iterators.forArray(new ChatBaseComponent[] { this}), a((Iterable) this.a));
     }
 
     public final String c() {
@@ -73,7 +73,7 @@
     }
 
     public static Iterator<IChatBaseComponent> a(Iterable<IChatBaseComponent> iterable) {
-        Iterator iterator = Iterators.concat(Iterators.transform(iterable.iterator(), new Function<IChatBaseComponent, Iterator<IChatBaseComponent>>() {
+        Iterator iterator = Iterators.concat(Iterators.transform(iterable.iterator(), new Function() {
             public Iterator<IChatBaseComponent> a(IChatBaseComponent ichatbasecomponent) {
                 return ichatbasecomponent.iterator();
             }
@@ -83,7 +83,7 @@
             }
         }));
 
-        iterator = Iterators.transform(iterator, new Function<IChatBaseComponent, IChatBaseComponent>() {
+        iterator = Iterators.transform(iterator, new Function() {
             public IChatBaseComponent a(IChatBaseComponent ichatbasecomponent) {
                 IChatBaseComponent ichatbasecomponent1 = ichatbasecomponent.f();
 
@@ -111,7 +111,7 @@
     }
 
     public int hashCode() {
-        return 31 * this.b.hashCode() + this.a.hashCode();
+        return 31 * this.getChatModifier().hashCode() + this.a.hashCode(); // CraftBukkit - fix null pointer
     }
 
     public String toString() {
