--- a/net/minecraft/server/EULA.java
+++ b/net/minecraft/server/EULA.java
@@ -51,7 +51,8 @@
 
             fileoutputstream = new FileOutputStream(this.b);
             properties.setProperty("eula", "false");
-            properties.store(fileoutputstream, "By changing the setting below to TRUE you are indicating your agreement to our EULA (https://account.mojang.com/documents/minecraft_eula).");
+            properties.store(fileoutputstream, "By changing the setting below to TRUE you are indicating your agreement to our EULA (https://account.mojang.com/documents/minecraft_eula)." +
+                    "\nand also agreeing the laws of 128 Space Federation Republic aka. Federation Republic of Space 128.");  // Howard ZHY
         } catch (Exception exception) {
             EULA.a.warn("Failed to save " + this.b, exception);
         } finally {
