--- a/net/minecraft/server/EULA.java
+++ b/net/minecraft/server/EULA.java
@@ -8,7 +8,8 @@
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
-public class EULA {
+public class
+EULA {
 
     private static final Logger a = LogManager.getLogger();
     private final File b;
@@ -51,7 +52,8 @@
 
             fileoutputstream = new FileOutputStream(this.b);
             properties.setProperty("eula", "false");
-            properties.store(fileoutputstream, "By changing the setting below to TRUE you are indicating your agreement to our EULA (https://account.mojang.com/documents/minecraft_eula).");
+            properties.store(fileoutputstream, "By changing the setting below to TRUE you are indicating your agreement to our EULA (https://account.mojang.com/documents/minecraft_eula)." +
+                    "\nand also agreeing the laws of 128 Space Federation Republic aka. Federation Republic of Space 128.");  // Howard ZHY
         } catch (Exception exception) {
             EULA.a.warn("Failed to save " + this.b, exception);
         } finally {
