--- a/net/minecraft/server/ItemMilkBucket.java
+++ b/net/minecraft/server/ItemMilkBucket.java
@@ -1,5 +1,7 @@
 package net.minecraft.server;
 
+import org.github.paperspigot.PaperSpigotConfig; // PaperSpigot
+
 public class ItemMilkBucket extends Item {
 
     public ItemMilkBucket() {
@@ -17,6 +19,15 @@
         }
 
         entityhuman.b(StatisticList.USE_ITEM_COUNT[Item.getId(this)]);
+        // PaperSpigot start - Stackable Buckets
+        if (PaperSpigotConfig.stackableMilkBuckets) {
+            if (itemstack.count <= 0) {
+                return new ItemStack(Items.BUCKET);
+            } else if (!entityhuman.inventory.pickup(new ItemStack(Items.BUCKET))) {
+                entityhuman.drop(new ItemStack(Items.BUCKET), false);
+            }
+        }
+        // PaperSpigot end
         return itemstack.count <= 0 ? new ItemStack(Items.BUCKET) : itemstack;
     }
 
