--- a/net/minecraft/server/ItemSkull.java
+++ b/net/minecraft/server/ItemSkull.java
@@ -5,7 +5,7 @@
 
 public class ItemSkull extends Item {
 
-    private static final String[] a = new String[]{"skeleton", "wither", "zombie", "char", "creeper"};
+    private static final String[] a = new String[] { "skeleton", "wither", "zombie", "char", "creeper"};
 
     public ItemSkull() {
         this.a(CreativeModeTab.c);
@@ -35,6 +35,12 @@
                 return false;
             } else {
                 if (!world.isClientSide) {
+                    // Spigot Start
+                    if ( !Blocks.SKULL.canPlace( world, blockposition ) )
+                    {
+                        return false;
+                    }
+                    // Spigot End
                     world.setTypeAndData(blockposition, Blocks.SKULL.getBlockData().set(BlockSkull.FACING, enumdirection), 3);
                     int i = 0;
 
@@ -94,14 +100,14 @@
     public String a(ItemStack itemstack) {
         if (itemstack.getData() == 3 && itemstack.hasTag()) {
             if (itemstack.getTag().hasKeyOfType("SkullOwner", 8)) {
-                return LocaleI18n.a("item.skull.player.name", itemstack.getTag().getString("SkullOwner"));
+                return LocaleI18n.a("item.skull.player.name", new Object[] { itemstack.getTag().getString("SkullOwner")});
             }
 
             if (itemstack.getTag().hasKeyOfType("SkullOwner", 10)) {
                 NBTTagCompound nbttagcompound = itemstack.getTag().getCompound("SkullOwner");
 
                 if (nbttagcompound.hasKeyOfType("Name", 8)) {
-                    return LocaleI18n.a("item.skull.player.name", nbttagcompound.getString("Name"));
+                    return LocaleI18n.a("item.skull.player.name", new Object[] { nbttagcompound.getString("Name")});
                 }
             }
         }
@@ -109,13 +115,21 @@
         return super.a(itemstack);
     }
 
-    public boolean a(NBTTagCompound nbttagcompound) {
+    public boolean a(final NBTTagCompound nbttagcompound) { // Spigot - make final
         super.a(nbttagcompound);
         if (nbttagcompound.hasKeyOfType("SkullOwner", 8) && nbttagcompound.getString("SkullOwner").length() > 0) {
             GameProfile gameprofile = new GameProfile((UUID) null, nbttagcompound.getString("SkullOwner"));
 
-            gameprofile = TileEntitySkull.b(gameprofile);
-            nbttagcompound.set("SkullOwner", GameProfileSerializer.serialize(new NBTTagCompound(), gameprofile));
+            // Spigot start
+            TileEntitySkull.b(gameprofile, new com.google.common.base.Predicate<GameProfile>() {
+
+                @Override
+                public boolean apply(GameProfile gameprofile) {
+                    nbttagcompound.set("SkullOwner", GameProfileSerializer.serialize(new NBTTagCompound(), gameprofile));                    
+                    return false;
+                }
+            });
+            // Spigot end
             return true;
         } else {
             return false;
