--- a/net/minecraft/server/NBTTagIntArray.java
+++ b/net/minecraft/server/NBTTagIntArray.java
@@ -27,6 +27,7 @@
     void load(DataInput datainput, int i, NBTReadLimiter nbtreadlimiter) throws IOException {
         nbtreadlimiter.a(192L);
         int j = datainput.readInt();
+       com.google.common.base.Preconditions.checkArgument( j < 1 << 24);
 
         nbtreadlimiter.a((long) (32 * j));
         this.data = new int[j];
@@ -38,7 +39,7 @@
     }
 
     public byte getTypeId() {
-        return 11;
+        return (byte) 11;
     }
 
     public String toString() {
