--- a/net/minecraft/server/OldChunkLoader.java
+++ b/net/minecraft/server/OldChunkLoader.java
@@ -46,15 +46,15 @@
         for (int l = 0; l < 8; ++l) {
             boolean flag = true;
 
-            for (k = 0; k < 16 && flag; ++k) {
-                j = 0;
+            for (j = 0; j < 16 && flag; ++j) {
+                k = 0;
 
-                while (j < 16 && flag) {
+                while (k < 16 && flag) {
                     int i1 = 0;
 
                     while (true) {
                         if (i1 < 16) {
-                            int j1 = k << 11 | i1 << 7 | j + (l << 4);
+                            int j1 = j << 11 | i1 << 7 | k + (l << 4);
                             byte b0 = oldchunkloader_oldchunk.g[j1];
 
                             if (b0 == 0) {
@@ -65,7 +65,7 @@
                             flag = false;
                         }
 
-                        ++j;
+                        ++k;
                         break;
                     }
                 }
