--- a/net/minecraft/server/PacketPlayInArmAnimation.java
+++ b/net/minecraft/server/PacketPlayInArmAnimation.java
@@ -4,17 +4,17 @@
 
 public class PacketPlayInArmAnimation implements Packet<PacketListenerPlayIn> {
 
+    public long timestamp; // Spigot
+
     public PacketPlayInArmAnimation() {}
 
-    public void a(PacketDataSerializer packetdataserializer) throws IOException {}
+    public void a(PacketDataSerializer packetdataserializer) throws IOException {
+        timestamp = System.currentTimeMillis(); // Spigot
+    }
 
     public void b(PacketDataSerializer packetdataserializer) throws IOException {}
 
     public void a(PacketListenerPlayIn packetlistenerplayin) {
         packetlistenerplayin.a(this);
     }
-
-    public void a(PacketListener packetlistener) {
-        this.a((PacketListenerPlayIn) packetlistener);
-    }
 }
