--- a/net/minecraft/server/PacketPlayInChat.java
+++ b/net/minecraft/server/PacketPlayInChat.java
@@ -24,15 +24,28 @@
         packetdataserializer.a(this.a);
     }
 
-    public void a(PacketListenerPlayIn packetlistenerplayin) {
+    // Spigot Start
+    private static final java.util.concurrent.ExecutorService executors = java.util.concurrent.Executors.newCachedThreadPool(
+            new com.google.common.util.concurrent.ThreadFactoryBuilder().setDaemon( true ).setNameFormat( "Async Chat Thread - #%d" ).build() );
+    public void a(final PacketListenerPlayIn packetlistenerplayin) {
+        if ( !a.startsWith("/") )
+        {
+            executors.submit( new Runnable()
+            {
+
+                @Override
+                public void run()
+                {
+                    packetlistenerplayin.a( PacketPlayInChat.this );
+                }
+            } );
+            return;
+        }
+        // Spigot End
         packetlistenerplayin.a(this);
     }
 
     public String a() {
         return this.a;
     }
-
-    public void a(PacketListener packetlistener) {
-        this.a((PacketListenerPlayIn) packetlistener);
-    }
 }
