--- a/net/minecraft/server/PacketEncoder.java
+++ b/net/minecraft/server/PacketEncoder.java
@@ -27,11 +27,11 @@
         }
 
         if (integer == null) {
-            throw new IOException("Can't serialize unregistered packet");
+            throw new IOException("Can\'t serialize unregistered packet");
         } else {
             PacketDataSerializer packetdataserializer = new PacketDataSerializer(bytebuf);
 
-            packetdataserializer.b(integer);
+            packetdataserializer.b(integer.intValue());
 
             try {
                 if (packet instanceof PacketPlayOutNamedEntitySpawn) {
@@ -46,7 +46,8 @@
         }
     }
 
-    protected void encode(ChannelHandlerContext channelhandlercontext, Object object, ByteBuf bytebuf) throws Exception {
-        this.a(channelhandlercontext, (Packet) object, bytebuf);
+    @Override
+    protected void encode(ChannelHandlerContext channelHandlerContext, Packet packet, ByteBuf byteBuf) throws Exception {
+        this.a(channelHandlerContext, (Packet) packet, byteBuf);
     }
 }
