--- a/net/minecraft/server/PacketPlayInCustomPayload.java
+++ b/net/minecraft/server/PacketPlayInCustomPayload.java
@@ -1,44 +1,41 @@
 package net.minecraft.server;
 
-import io.netty.buffer.ByteBuf;
+import org.bukkit.Bukkit;
+
 import java.io.IOException;
 
 public class PacketPlayInCustomPayload implements Packet<PacketListenerPlayIn> {
+  // BurritoSpigot Start - PAIL
+  public String tag = this.a;
 
-    private String a;
-    private PacketDataSerializer b;
-
-    public PacketPlayInCustomPayload() {}
-
-    public void a(PacketDataSerializer packetdataserializer) throws IOException {
-        this.a = packetdataserializer.c(20);
-        int i = packetdataserializer.readableBytes();
-
-        if (i >= 0 && i <= 32767) {
-            this.b = new PacketDataSerializer(packetdataserializer.readBytes(i));
-        } else {
-            throw new IOException("Payload may not be larger than 32767 bytes");
-        }
-    }
-
-    public void b(PacketDataSerializer packetdataserializer) throws IOException {
-        packetdataserializer.a(this.a);
-        packetdataserializer.writeBytes((ByteBuf) this.b);
-    }
-
-    public void a(PacketListenerPlayIn packetlistenerplayin) {
-        packetlistenerplayin.a(this);
-    }
-
-    public String a() {
-        return this.a;
-    }
-
-    public PacketDataSerializer b() {
-        return this.b;
-    }
-
-    public void a(PacketListener packetlistener) {
-        this.a((PacketListenerPlayIn) packetlistener);
-    }
-}
+  public String a;
+  
+  public PacketDataSerializer b;
+  // BurritoSpigot End
+
+  public void a(PacketDataSerializer paramPacketDataSerializer) throws IOException
+  {
+    this.a = paramPacketDataSerializer.c(20);
+    int i = paramPacketDataSerializer.readableBytes();
+    if (i < 0 || i > 32767)
+      throw new IOException("Payload may not be larger than 32767 bytes"); 
+    this.b = new PacketDataSerializer(paramPacketDataSerializer.readBytes(i));
+  }
+  
+  public void b(PacketDataSerializer paramPacketDataSerializer) throws IOException {
+    paramPacketDataSerializer.a(this.a);
+    paramPacketDataSerializer.writeBytes(this.b);
+  }
+  
+  public void a(PacketListenerPlayIn paramPacketListenerPlayIn) {
+    paramPacketListenerPlayIn.a(this);
+  }
+  
+  public String a() {
+    return this.a;
+  }
+  
+  public PacketDataSerializer b() {
+    return this.b;
+  }
+}
\ No newline at end of file
