--- a/net/minecraft/server/PathfinderGoalNearestAttackableTargetInsentient.java
+++ b/net/minecraft/server/PathfinderGoalNearestAttackableTargetInsentient.java
@@ -22,7 +22,7 @@
             PathfinderGoalNearestAttackableTargetInsentient.a.warn("Use NearestAttackableTargetGoal.class for PathfinerMob mobs!");
         }
 
-        this.c = new Predicate<EntityLiving>() {
+        this.c = new Predicate() {
             public boolean a(EntityLiving entityliving) {
                 double d0 = PathfinderGoalNearestAttackableTargetInsentient.this.f();
 
@@ -68,7 +68,7 @@
     }
 
     public void c() {
-        this.b.setGoalTarget(this.e);
+        this.b.setGoalTarget(this.e, org.bukkit.event.entity.EntityTargetEvent.TargetReason.CLOSEST_ENTITY, true); // CraftBukkit - reason
         super.c();
     }
 
