--- a/net/minecraft/server/RegionFileCache.java
+++ b/net/minecraft/server/RegionFileCache.java
@@ -10,9 +10,14 @@
 
 public class RegionFileCache {
 
-    private static final Map<File, RegionFile> a = Maps.newHashMap();
+    public static final Map<File, RegionFile> a = Maps.newHashMap(); // Spigot - private -> public
 
+    // PaperSpigot start
     public static synchronized RegionFile a(File file, int i, int j) {
+        return a(file, i, j, true);
+    }
+    public static synchronized RegionFile a(File file, int i, int j, boolean create) {
+        // PaperSpigot end
         File file1 = new File(file, "region");
         File file2 = new File(file1, "r." + (i >> 5) + "." + (j >> 5) + ".mca");
         RegionFile regionfile = (RegionFile) RegionFileCache.a.get(file2);
@@ -20,6 +25,7 @@
         if (regionfile != null) {
             return regionfile;
         } else {
+            if (!create && !file2.exists()) { return null; } // PaperSpigot
             if (!file1.exists()) {
                 file1.mkdirs();
             }
