--- a/net/minecraft/server/WorldGenForestTree.java
+++ b/net/minecraft/server/WorldGenForestTree.java
@@ -5,7 +5,7 @@
 public class WorldGenForestTree extends WorldGenTreeAbstract {
 
     private static final IBlockData a = Blocks.LOG2.getBlockData().set(BlockLog2.VARIANT, BlockWood.EnumLogVariant.DARK_OAK);
-    private static final IBlockData b = Blocks.LEAVES2.getBlockData().set(BlockLeaves2.VARIANT, BlockWood.EnumLogVariant.DARK_OAK).set(BlockLeaves.CHECK_DECAY, false);
+    private static final IBlockData b = Blocks.LEAVES2.getBlockData().set(BlockLeaves2.VARIANT, BlockWood.EnumLogVariant.DARK_OAK).set(BlockLeaves.CHECK_DECAY, Boolean.valueOf(false));
 
     public WorldGenForestTree(boolean flag) {
         super(flag);
